src/cpu/pred/simple_btb.cc

```diff
@@ -0,0 +1,181 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "cpu/pred/simple_btb.hh"
+
+#include "base/intmath.hh"
+#include "base/trace.hh"
+#include "debug/BTB.hh"
+
+namespace gem5
+{
+
+namespace branch_prediction
+{
+
+SimpleBTB::SimpleBTB(const SimpleBTBParams &p)
+    : BranchTargetBuffer(p),
+        numEntries(p.numEntries),
+        tagBits(p.tagBits),
+        instShiftAmt(p.instShiftAmt),
+        log2NumThreads(floorLog2(p.numThreads))
+{
+    DPRINTF(BTB, "BTB: Creating BTB object.\n");
+
+    if (!isPowerOf2(numEntries)) {
+        fatal("BTB entries is not a power of 2!");
+    }
+
+    btb.resize(numEntries);
+
+    for (unsigned i = 0; i < numEntries; ++i) {
+        btb[i].valid = false;
+    }
+
+    idxMask = numEntries - 1;
+
+    tagMask = (1 << tagBits) - 1;
+
+    tagShiftAmt = instShiftAmt + floorLog2(numEntries);
+}
+
+void
+SimpleBTB::memInvalidate()
+{
+    for (unsigned i = 0; i < numEntries; ++i) {
+        btb[i].valid = false;
+    }
+}
+
+inline
+unsigned
+SimpleBTB::getIndex(Addr instPC, ThreadID tid)
+{
+    // Need to shift PC over by the word offset.
+    return ((instPC >> instShiftAmt)
+            ^ (tid << (tagShiftAmt - instShiftAmt - log2NumThreads)))
+            & idxMask;
+}
+
+inline
+Addr
+SimpleBTB::getTag(Addr instPC)
+{
+    return (instPC >> tagShiftAmt) & tagMask;
+}
+
+bool
+SimpleBTB::valid(ThreadID tid, Addr instPC)
+{
+    unsigned btb_idx = getIndex(instPC, tid);
+
+    Addr inst_tag = getTag(instPC);
+
+    assert(btb_idx < numEntries);
+
+    if (btb[btb_idx].valid
+        && inst_tag == btb[btb_idx].tag
+        && btb[btb_idx].tid == tid) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+// @todo Create some sort of return struct that has both whether or not the
+// address is valid, and also the address.  For now will just use addr = 0 to
+// represent invalid entry.
+const PCStateBase *
+SimpleBTB::lookup(ThreadID tid, Addr instPC, BranchType type)
+{
+    unsigned btb_idx = getIndex(instPC, tid);
+    Addr inst_tag = getTag(instPC);
+
+    assert(btb_idx < numEntries);
+
+    stats.lookups[type]++;
+
+    if (btb[btb_idx].valid
+        && inst_tag == btb[btb_idx].tag
+        && btb[btb_idx].tid == tid) {
+        return btb[btb_idx].target.get();
+    } else {
+        stats.misses[type]++;
+        return nullptr;
+    }
+}
+
+const StaticInstPtr
+SimpleBTB::lookupInst(ThreadID tid, Addr instPC)
```

Comment:
There is a lot of duplication between the two lookup methods. I'd suggest refactoring this so you have a `getBTBEntry` (or something along those lines) method that returns a `BTBEntry` or `nullptr` that you use in both.

Reply:
Great suggestion! Thanks for that.
I combined now all lookup functions in an internal function.